cpp/src/parquet/encoding.cc

```diff
@@ -1429,8 +1429,15 @@ class PlainByteArrayDecoder : public PlainDecoder<ByteArrayType>,
         std::min<int64_t>(len_, helper.chunk_space_remaining)));
     for (int i = 0; i < num_values; ++i) {
       if (bit_reader.IsSet()) {
-        auto value_len = static_cast<int32_t>(arrow::util::SafeLoadAs<uint32_t>(data_));
-        int increment = static_cast<int>(sizeof(uint32_t) + value_len);
+        // For compiler warnings on unsigned/signed arithmetic.
+        auto int32_s = static_cast<int32_t>(sizeof(int32_t));
+        auto value_len = arrow::util::SafeLoadAs<int32_t>(data_);
+        auto increment = int32_s + value_len;
+
+        if (ARROW_PREDICT_FALSE(value_len < 0 || value_len > INT32_MAX - int32_s)) {
```

Comment:
You want to do this before computing `increment`, no? Otherwise, you can get UB...